src/k8s/pkg/k8sd/controllers/control_plane_configuration_test.go

```diff
@@ -0,0 +1,319 @@
+
+				select {
+				case triggerCh <- time.Now():
+				case <-time.After(channelSendTimeout):
+					g.Fail("Timed out while attempting to trigger controller reconcile loop")
+				}
+
+				// TODO: this should be changed to call g.Eventually()
+				<-time.After(50 * time.Millisecond)
+
+				g.Expect(s.RestartServiceCalledWith).To(ConsistOf(tc.expectServiceRestarts))
+
+				t.Run("APIServerArgs", func(t *testing.T) {
+					for earg, eval := range tc.expectKubeAPIServerArgs {
+						t.Run(earg, func(t *testing.T) {
+							g := NewWithT(t)
+
+							val, err := snaputil.GetServiceArgument(s, "kube-apiserver", earg)
+							g.Expect(err).To(BeNil())
+							g.Expect(val).To(Equal(eval))
+						})
+					}
+				})
+
+				t.Run("KubeControllerManagerArgs", func(t *testing.T) {
+					for earg, eval := range tc.expectKubeControllerManagerArgs {
+						t.Run(earg, func(t *testing.T) {
+							g := NewWithT(t)
+
+							val, err := snaputil.GetServiceArgument(s, "kube-controller-manager", earg)
+							g.Expect(err).To(BeNil())
+							g.Expect(val).To(Equal(eval))
+						})
+					}
+				})
+
+				t.Run("Certs", func(t *testing.T) {
+					for file, mustExist := range tc.expectFilesToExist {
+						t.Run(path.Base(file), func(t *testing.T) {
+							g := NewWithT(t)
+
+							_, err := os.Stat(file)
+							if mustExist {
+								g.Expect(err).To(BeNil())
+							} else {
+								g.Expect(err).To(MatchError(os.ErrNotExist))
+							}
+						})
+					}
+				})
+			})
+		}
+	})
+
+	t.Run("Worker", func(t *testing.T) {
+		dir := t.TempDir()
+
+		s := &mock.Snap{
+			Mock: mock.Mock{
+				EtcdPKIDir:          path.Join(dir, "etcd-pki"),
+				ServiceArgumentsDir: path.Join(dir, "args"),
+				LockFilesDir:        path.Join(dir, "locks"),
+				UID:                 os.Getuid(),
+				GID:                 os.Getgid(),
+			},
+		}
+
+		g := NewWithT(t)
+		g.Expect(setup.EnsureAllDirectories(s)).To(Succeed())
+
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+
+		triggerCh := make(chan time.Time)
+		configProvider := &configProvider{}
+
+		ctrl := controllers.NewControlPlaneConfigurationController(s, func() {}, triggerCh)
+		go ctrl.Run(ctx, configProvider.getConfig)
+
+		// mark as worker node
+		g.Expect(snaputil.MarkAsWorkerNode(s, true)).To(Succeed())
+
+		configProvider.config = types.ClusterConfig{
+			Datastore: types.Datastore{
+				Type:               vals.Pointer("external"),
+				ExternalURL:        vals.Pointer("https://127.0.0.1:2379"),
+				ExternalCACert:     vals.Pointer("CA DATA"),
+				ExternalClientCert: vals.Pointer("CERT DATA"),
+				ExternalClientKey:  vals.Pointer("KEY DATA"),
+			},
+			Kubelet: types.Kubelet{
+				CloudProvider: vals.Pointer("external"),
+			},
+		}
+
+		select {
+		case triggerCh <- time.Now():
+		case <-time.After(channelSendTimeout):
+			g.Fail("Timed out while attempting to trigger controller reconcile loop")
+		}
+
+		// TODO: this should be changed to call g.Eventually()
```

Comment:
nit: what is currently blocking this?

Reply:
it's just about spending proper effort to rewrite the tests accordingly.

Reply:
can be done at a later time, not pressing atm